instrumentation/aws_lambda/lib/opentelemetry/instrumentation/aws_lambda/wrap.rb

```diff
@@ -0,0 +1,171 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module AwsLambda
+      # Helper module that can be used to wrap a lambda handler method
+      module Wrap
+        AWS_TRIGGERS = ['aws:sqs', 'aws:s3', 'aws:sns', 'aws:dynamodb'].freeze
+        DEFAULT_FLUSH_TIMEOUT = ENV.fetch('OTEL_INSTRUMENTATION_AWS_LAMBDA_FLUSH_TIMEOUT', '30000').to_i
+
+        def instrument_handler(method, flush_timeout: DEFAULT_FLUSH_TIMEOUT)
+          raise ArgumentError, "#{method} is not a method of #{name}" unless respond_to?(method)
+
+          uninstrumented_method = "#{method}_without_instrumentation"
+          singleton_class.alias_method uninstrumented_method, method
+
+          handler = "#{name}.#{method}"
+
+          define_singleton_method(method) do |event:, context:|
+            wrap_lambda(event:, context:, handler:, flush_timeout:) { public_send(uninstrumented_method, event:, context:) }
+          end
+        end
+
+        # Try to record and re-raise any exception from the wrapped function handler
+        # Instrumentation should never raise its own exception
+        def wrap_lambda(event:, context:, handler:, flush_timeout: DEFAULT_FLUSH_TIMEOUT)
+          parent_context = extract_parent_context(event)
+
+          span_kind = if event['Records'] && AWS_TRIGGERS.include?(event['Records'].dig(0, 'eventSource'))
+                        :consumer
+                      else
+                        :server
+                      end
+
+          original_handler_error = nil
+          original_response = nil
+          OpenTelemetry::Context.with_current(parent_context) do
+            tracer.in_span(handler, attributes: otel_attributes(event, context), kind: span_kind) do |span|
+              begin
+                response = yield
+
+                unless span.attributes.key?(OpenTelemetry::SemanticConventions::Trace::HTTP_STATUS_CODE)
+                  status_code = response['statusCode'] || response[:statusCode] if response.is_a?(Hash)
+                  span.set_attribute(OpenTelemetry::SemanticConventions::Trace::HTTP_STATUS_CODE, status_code) if status_code
+                end
+              rescue StandardError => e
+                original_handler_error = e
+              ensure
+                original_response = response
+              end
+              if original_handler_error
+                span.record_exception(original_handler_error)
+                span.status = OpenTelemetry::Trace::Status.error(original_handler_error.message)
+              end
+            end
+          end
+
+          OpenTelemetry.tracer_provider.force_flush(timeout: flush_timeout)
+          OpenTelemetry.meter_provider.force_flush(timeout: flush_timeout)
```

Comment:
Probably need to throw some `try`s in here so we don't blow up if the metrics SDK isn't in use